s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/input/ByteArrayTransformer.java

```diff
@@ -14,39 +14,56 @@
  * limitations under the License.
  */
 
-package io.aiven.kafka.connect.s3.source.output;
+package io.aiven.kafka.connect.s3.source.input;
+
+import static io.aiven.kafka.connect.s3.source.config.S3SourceConfig.EXPECTED_MAX_MESSAGE_BYTES;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 
 import io.aiven.kafka.connect.s3.source.config.S3SourceConfig;
 
-import com.amazonaws.util.IOUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class ByteArrayWriter implements OutputWriter {
-    private static final Logger LOGGER = LoggerFactory.getLogger(ByteArrayWriter.class);
+public class ByteArrayTransformer implements Transformer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ByteArrayTransformer.class);
 
     @Override
     public void configureValueConverter(final Map<String, String> config, final S3SourceConfig s3SourceConfig) {
-
+        // For byte array transformations, ByteArrayConverter is the converter which is the default config.
     }
 
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
     @Override
-    public List<Object> getRecords(final InputStream inputStream, final String topic, final int topicPartition) {
-        return List.of(inputStream);
-    }
+    public List<Object> getRecords(final InputStream inputStream, final String topic, final int topicPartition,
+            final S3SourceConfig s3SourceConfig) {
 
-    @Override
-    public byte[] getValueBytes(final Object record, final String topic, final S3SourceConfig s3SourceConfig) {
+        final int maxMessageBytesSize = s3SourceConfig.getInt(EXPECTED_MAX_MESSAGE_BYTES);
+        final byte[] buffer = new byte[maxMessageBytesSize];
+        int bytesRead;
+
+        final List<Object> chunks = new ArrayList<>();
         try {
-            return IOUtils.toByteArray((InputStream) record);
+            bytesRead = inputStream.read(buffer);
+            while (bytesRead != -1) {
+                final byte[] chunk = new byte[bytesRead];
+                System.arraycopy(buffer, 0, chunk, 0, bytesRead);
+                chunks.add(chunk);
+                bytesRead = inputStream.read(buffer);
+            }
```

Comment:
Here AFAIU we read the whole inputStream that is a whole S3 object, we should not do this.

Reply:
Let's discuss this during refinement. Probably we need to apply this on all transformers. Ticket https://aiven.atlassian.net/browse/KCON-35